### PR TITLE
pybind/mgr/balancer: fix pool-deletion vs auto-optimization race

### PR DESCRIPTION
**Pull request by xiexingguo**
This patch fixes the error below:
```
File "/usr/lib/ceph/mgr/balancer/module.py", line 722, in optimize
  return self.do_crush_compat(plan)
File "/usr/lib/ceph/mgr/balancer/module.py", line 781, in do_crush_compat
  pe = self.calc_eval(ms, plan.pools)
File "/usr/lib/ceph/mgr/balancer/module.py", line 570, in calc_eval
  objects_by_osd[osd] += ms.pg_stat[pgid]['num_objects']
KeyError: ('5.1b',)
```

The root cause is that balancer is basically collecting cluster
information from two separate maps (OSDMap and PGMap), and hence
there is a small window/chance that the pool statistics might
become divergent. E.g.:
1) auto-optimization begin
2) get osdmap
3) a pool is gone (deleted by admin); pg_dump refreshed
4) get pg_dump (balancer is now with both the newest pg_dump
   and an obsolute osdmap in hand)
5) execute optimization; balancer complains some PGs are missing
   in the pg_dump map..

Fix the above problem by tracing pools existing in both maps only.

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

File: src/pybind/mgr/balancer/module.py [hunk 1]
```diff
@@ -31,7 +31,9 @@ def __init__(self, osdmap, pg_dump, desc=''):
         self.pg_stat = {
             i['pgid']: i['stat_sum'] for i in pg_dump.get('pg_stats', [])
         }
-        self.poolids = [p['pool'] for p in self.osdmap_dump.get('pools', [])]
+        osd_poolids = [p['pool'] for p in self.osdmap_dump.get('pools', [])]
+        pg_poolids = [p['poolid'] for p in pg_dump.get('pool_stats', [])]
+        self.poolids = [p for p in osd_poolids if p in pg_poolids]
```
Comment (jcsp) [inline src/pybind/mgr/balancer/module.py]: Would be a bit neater (and perhaps more efficient) to do a `set(osd_poolids) & set(pg_poolids)`

Comment (xiexingguo) [reply src/pybind/mgr/balancer/module.py]: This is still a work-in-progress and I have some other bugs to hunt :joy: 
Will fix in next version \o/

**Comment by xiexingguo** [comment]
@liewegas I hit another balancer issue. Mind taking a look at this one too? Thanks!

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-08_12:59:37-rados-wip-kefu-testing-2018-03-08-1932-distro-basic-smithi/

- the swift failure was caused by #20419
- and i reran `osd-pool-create.sh` multiple times locally, and was not able to reproduce that failure.

**Comment by xiexingguo** [comment]
Thanks @tchaikov !

**Pull request merged**